src/parse.c

```diff
@@ -1648,6 +1648,184 @@ handle_tunnel_key_mapping(yaml_document_t* doc, yaml_node_t* node, const void* _
+    }
+    /* split off :port */
+    if (!port)
+        return yaml_error(node, error, "endpoint '%s' is missing :port", scalar(node));
+    *port = '\0';
+    port++; /* skip former ':' into first char of port */
+    port_num = g_ascii_strtoull(port, NULL, 10);
+    if (port_num > 65535)
+        return yaml_error(node, error, "invalid port in endpoint '%s'", scalar(node));
+    if (is_ip4_address(address) || is_ip6_address(address) || is_hostname(address)) {
+        return handle_wireguard_peer_str(doc, node, &peer->endpoint, error);
+    }
+    return yaml_error(node, error, "invalid endpoint address or hostname '%s'", scalar(node));
+}
+
+#define wireguard_peer_offset(field) GUINT_TO_POINTER(offsetof(wireguard_peer, field))
+const mapping_entry_handler wireguard_peer_handlers[] = {
+    {"public-key", YAML_SCALAR_NODE, handle_wireguard_peer_str, NULL, wireguard_peer_offset(public_key)},
+    {"preshared-key", YAML_SCALAR_NODE, handle_wireguard_peer_str, NULL, wireguard_peer_offset(preshared_key)},
+    {"preshared-key-file", YAML_SCALAR_NODE, handle_wireguard_peer_str, NULL, wireguard_peer_offset(preshared_key_file)},
+    {"keepalive", YAML_SCALAR_NODE, handle_wireguard_peer_guint, NULL, wireguard_peer_offset(keepalive)},
+    {"endpoint", YAML_SCALAR_NODE, handle_wireguard_endpoint},
+    {"allowed-ips", YAML_SEQUENCE_NODE, handle_wireguard_allowed_ips},
+
+    {NULL}
+};
+
+static gboolean
+process_wireguard_peer(yaml_document_t* doc, yaml_node_t* node, wireguard_peer *peer, GError** error)
+{
+    for (yaml_node_pair_t *entry = node->data.mapping.pairs.start; entry < node->data.mapping.pairs.top; entry++) {
+        yaml_node_t* key, *value;
+        const mapping_entry_handler* h;
+
+        g_assert(*error == NULL);
+
+        key = yaml_document_get_node(doc, entry->key);
+        value = yaml_document_get_node(doc, entry->value);
+        assert_type(key, YAML_SCALAR_NODE);
+        h = get_handler(wireguard_peer_handlers, scalar(key));
+        if (!h)
+            return yaml_error(key, error, "unknown key '%s'", scalar(key));
+        assert_type(value, h->type);
+
+        if (!h->handler(doc, value, (void *)peer + GPOINTER_TO_UINT(h->data), error))
+            return FALSE;
+    }
+    return TRUE;
+}
+
+static gboolean
+handle_wireguard_peers(yaml_document_t* doc, yaml_node_t* node, const void* _, GError** error)
+{
+    cur_netdef->wireguard_peers = g_array_new(FALSE, TRUE, sizeof(wireguard_peer));
+    for (yaml_node_item_t *i = node->data.sequence.items.start; i < node->data.sequence.items.top; i++) {
+        g_autofree char* addr = NULL;
+        yaml_node_t *entry = yaml_document_get_node(doc, *i);
+        assert_type(entry, YAML_MAPPING_NODE);
+
+        wireguard_peer peer = {0};
+        peer.allowed_ips = g_array_new(FALSE, FALSE, sizeof(char*));
+
+        gboolean ret = process_wireguard_peer(doc, entry, &peer, error);
+        if (ret) {
+            g_array_append_val(cur_netdef->wireguard_peers, peer);
```

Comment:
Please make sure not to insert the same peer multiple times, if the code is run through a 2nd pass. Similar to how I just recently fixed the same issue for IP4/IP6 addresses: https://github.com/CanonicalLtd/netplan/pull/147/commits/2f2971b7935c9ff6ae14cd68907cfb87b68472c4